### ISSUE TITLE
java.lang.NoClassDefFoundError: Could not initialize class com.amazonaws.ClientConfiguration

### ISSUE THREAD

**Issue by pmvilaca** [databricks/spark-redshift#133]
I'm trying to use this library but I'm not able to solve a problem related with aws sdk.

I'm running spark shell and adding the dependencies that are required but I'm getting the following stacktrace when I try to use it.

```
bin/spark-shell --packages com.databricks:spark-redshift_2.10:0.5.2,com.amazonaws:aws-java-sdk-s3:1.10.38,com.amazonaws:aws-java-sdk-core:1.10.38 --repositories http://repohost/content/groups/public --exclude-packages com.fasterxml.jackson.core:jackson-databind --jars ~/development/redshift/RedshiftJDBC41-1.1.10.1010.jar
```
Already inside the spark-shell:

```
scala> sqlContext.read.format("com.databricks.spark.redshift").option("url", "jdbc:redshift://redshifthost:5439/db?user=user&password=somepass").option("dbtable", "schema.table").option("tempdir", "s3n://s3bucket/spark-redshift-test").load()
res0: org.apache.spark.sql.DataFrame = [id: int, user_id: int, ...]
scala> res0.show
java.lang.NoClassDefFoundError: Could not initialize class com.amazonaws.ClientConfiguration
	at com.amazonaws.services.s3.AmazonS3Client.<init>(AmazonS3Client.java:389)
	at com.databricks.spark.redshift.DefaultSource$$anonfun$$init$$1.apply(DefaultSource.scala:39)
	at com.databricks.spark.redshift.DefaultSource$$anonfun$$init$$1.apply(DefaultSource.scala:39)
	at com.databricks.spark.redshift.RedshiftRelation.buildScan(RedshiftRelation.scala:83)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$3.apply(DataSourceStrategy.scala:53)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$3.apply(DataSourceStrategy.scala:53)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$pruneFilterProject$1.apply(DataSourceStrategy.scala:287)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$pruneFilterProject$1.apply(DataSourceStrategy.scala:286)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$.pruneFilterProjectRaw(DataSourceStrategy.scala:318)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$.pruneFilterProject(DataSourceStrategy.scala:282)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$.apply(DataSourceStrategy.scala:49)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:58)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:58)
	at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:371)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner.plan(QueryPlanner.scala:59)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner.planLater(QueryPlanner.scala:54)
	at org.apache.spark.sql.execution.SparkStrategies$BasicOperators$.apply(SparkStrategies.scala:374)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:58)
```

**Comment by JoshRosen** [comment]
After searching Google for "java.lang.NoClassDefFoundError: Could not initialize class com.amazonaws.ClientConfiguration", it looks like this problem may be caused by a missing `aws-java-sdk-core` dependency. For example: http://stackoverflow.com/a/32193634/590203

Is this the complete stacktrace or is there also a "Caused by: " entry with additional information?

**Comment by pmvilaca** [comment]
Sorry for the delay @JoshRosen 

There is no "Caused by: ..." in the stacktrace.. here is the complete stacktrace:

```
scala> res0.show
java.lang.NoClassDefFoundError: Could not initialize class com.amazonaws.ClientConfiguration
	at com.amazonaws.services.s3.AmazonS3Client.<init>(AmazonS3Client.java:395)
	at com.databricks.spark.redshift.DefaultSource$$anonfun$$init$$1.apply(DefaultSource.scala:39)
	at com.databricks.spark.redshift.DefaultSource$$anonfun$$init$$1.apply(DefaultSource.scala:39)
	at com.databricks.spark.redshift.RedshiftRelation.buildScan(RedshiftRelation.scala:83)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$3.apply(DataSourceStrategy.scala:53)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$3.apply(DataSourceStrategy.scala:53)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$pruneFilterProject$1.apply(DataSourceStrategy.scala:287)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$$anonfun$pruneFilterProject$1.apply(DataSourceStrategy.scala:286)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$.pruneFilterProjectRaw(DataSourceStrategy.scala:318)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$.pruneFilterProject(DataSourceStrategy.scala:282)
	at org.apache.spark.sql.execution.datasources.DataSourceStrategy$.apply(DataSourceStrategy.scala:49)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:58)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:58)
	at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:371)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner.plan(QueryPlanner.scala:59)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner.planLater(QueryPlanner.scala:54)
	at org.apache.spark.sql.execution.SparkStrategies$BasicOperators$.apply(SparkStrategies.scala:374)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:58)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner$$anonfun$1.apply(QueryPlanner.scala:58)
	at scala.collection.Iterator$$anon$13.hasNext(Iterator.scala:371)
	at org.apache.spark.sql.catalyst.planning.QueryPlanner.plan(QueryPlanner.scala:59)
	at org.apache.spark.sql.SQLContext$QueryExecution.sparkPlan$lzycompute(SQLContext.scala:926)
	at org.apache.spark.sql.SQLContext$QueryExecution.sparkPlan(SQLContext.scala:924)
	at org.apache.spark.sql.SQLContext$QueryExecution.executedPlan$lzycompute(SQLContext.scala:930)
	at org.apache.spark.sql.SQLContext$QueryExecution.executedPlan(SQLContext.scala:930)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:53)
	at org.apache.spark.sql.DataFrame.withNewExecutionId(DataFrame.scala:1903)
	at org.apache.spark.sql.DataFrame.collect(DataFrame.scala:1384)
	at org.apache.spark.sql.DataFrame.head(DataFrame.scala:1314)
	at org.apache.spark.sql.DataFrame.take(DataFrame.scala:1377)
	at org.apache.spark.sql.DataFrame.showString(DataFrame.scala:178)
	at org.apache.spark.sql.DataFrame.show(DataFrame.scala:401)
	at org.apache.spark.sql.DataFrame.show(DataFrame.scala:362)
	at org.apache.spark.sql.DataFrame.show(DataFrame.scala:370)
	at $iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC.<init>(<console>:22)
	at $iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC.<init>(<console>:27)
	at $iwC$$iwC$$iwC$$iwC$$iwC$$iwC.<init>(<console>:29)
	at $iwC$$iwC$$iwC$$iwC$$iwC.<init>(<console>:31)
	at $iwC$$iwC$$iwC$$iwC.<init>(<console>:33)
	at $iwC$$iwC$$iwC.<init>(<console>:35)
	at $iwC$$iwC.<init>(<console>:37)
	at $iwC.<init>(<console>:39)
	at <init>(<console>:41)
	at .<init>(<console>:45)
	at .<clinit>(<console>)
	at .<init>(<console>:7)
	at .<clinit>(<console>)
	at $print(<console>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.spark.repl.SparkIMain$ReadEvalPrint.call(SparkIMain.scala:1065)
	at org.apache.spark.repl.SparkIMain$Request.loadAndRun(SparkIMain.scala:1340)
	at org.apache.spark.repl.SparkIMain.loadAndRunReq$1(SparkIMain.scala:840)
	at org.apache.spark.repl.SparkIMain.interpret(SparkIMain.scala:871)
	at org.apache.spark.repl.SparkIMain.interpret(SparkIMain.scala:819)
	at org.apache.spark.repl.SparkILoop.reallyInterpret$1(SparkILoop.scala:857)
	at org.apache.spark.repl.SparkILoop.interpretStartingWith(SparkILoop.scala:902)
	at org.apache.spark.repl.SparkILoop.command(SparkILoop.scala:814)
	at org.apache.spark.repl.SparkILoop.processLine$1(SparkILoop.scala:657)
	at org.apache.spark.repl.SparkILoop.innerLoop$1(SparkILoop.scala:665)
	at org.apache.spark.repl.SparkILoop.org$apache$spark$repl$SparkILoop$$loop(SparkILoop.scala:670)
	at org.apache.spark.repl.SparkILoop$$anonfun$org$apache$spark$repl$SparkILoop$$process$1.apply$mcZ$sp(SparkILoop.scala:997)
	at org.apache.spark.repl.SparkILoop$$anonfun$org$apache$spark$repl$SparkILoop$$process$1.apply(SparkILoop.scala:945)
	at org.apache.spark.repl.SparkILoop$$anonfun$org$apache$spark$repl$SparkILoop$$process$1.apply(SparkILoop.scala:945)
	at scala.tools.nsc.util.ScalaClassLoader$.savingContextLoader(ScalaClassLoader.scala:135)
	at org.apache.spark.repl.SparkILoop.org$apache$spark$repl$SparkILoop$$process(SparkILoop.scala:945)
	at org.apache.spark.repl.SparkILoop.process(SparkILoop.scala:1059)
	at org.apache.spark.repl.Main$.main(Main.scala:31)
	at org.apache.spark.repl.Main.main(Main.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:674)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:180)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:205)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:120)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)


scala>
```

Regarding the missing ```aws-java-sdk-core```, it cannot be because I'm adding it to the classpath:

```
pedrovilaca@pvilaca-mbp spark-1.5.2-bin-hadoop2.4]$ ./bin/spark-shell --driver-class-path ~/development/redshift/RedshiftJDBC41-1.1.10.1010.jar --packages com.databricks:spark-redshift_2.10:0.5.2,com.amazonaws:aws-java-sdk-core:1.10.40,com.amazonaws:aws-java-sdk-s3:1.10.40 --exclude-packages com.fasterxml.jackson.core:jackson-databind
Ivy Default Cache set to: /Users/pedrovilaca/.ivy2/cache
The jars for the packages stored in: /Users/pedrovilaca/.ivy2/jars
:: loading settings :: url = jar:file:/Users/pedrovilaca/Downloads/spark-1.5.2-bin-hadoop2.4/lib/spark-assembly-1.5.2-hadoop2.4.0.jar!/org/apache/ivy/core/settings/ivysettings.xml
com.databricks#spark-redshift_2.10 added as a dependency
com.amazonaws#aws-java-sdk-core added as a dependency
com.amazonaws#aws-java-sdk-s3 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent;1.0
...
```

**Comment by pmvilaca** [comment]
I think that I found the problem.. I was using aws-java-sdk.core 1.10.40 and that was the reasonfor the error. The issue is solved if I use 1.10.22. 

Although, later I get the error that is described in #121

**Comment by JoshRosen** [comment]
To debug this a little further, I tried using the `spark-shell` listed in your last update, then ran `Class.forName("com.amazonaws.ClientConfiguration")` in `spark-shell`. It looks like this failed due to a binary incompatibility in Jackson:

```
./bin/spark-shell --driver-class-path ~/development/redshift/RedshiftJDBC41-1.1.10.1010.jar --packages com.databricks:spark-redshift_2.10:0.5.2,com.amazonaws:aws-java-sdk-core:1.10.40,com.amazonaws:aws-java-sdk-s3:1.10.40 --exclude-packages com.fasterxml.jackson.core:jackson-databind
```

```scala
scala> Class.forName("com.amazonaws.ClientConfiguration")
java.lang.NoSuchMethodError: com.fasterxml.jackson.databind.ObjectMapper.enable([Lcom/fasterxml/jackson/core/JsonParser$Feature;)Lcom/fasterxml/jackson/databind/ObjectMapper;
        at com.amazonaws.internal.config.InternalConfig.<clinit>(InternalConfig.java:43)
        at com.amazonaws.internal.config.InternalConfig$Factory.<clinit>(InternalConfig.java:304)
        at com.amazonaws.util.VersionInfoUtils.userAgent(VersionInfoUtils.java:139)
        at com.amazonaws.util.VersionInfoUtils.initializeUserAgent(VersionInfoUtils.java:134)
        at com.amazonaws.util.VersionInfoUtils.getUserAgent(VersionInfoUtils.java:95)
        at com.amazonaws.ClientConfiguration.<clinit>(ClientConfiguration.java:61)
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:264)
        at $iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC$$iwC.<init>(<console>:20)
[...]
```

Similarly, I get the same error when using aws-java-sdk 1.10.38.

I suspect that this is related to a conflict between Spark's version of Jackson and the version that the AWS SDK needs. It looks like this is a known issue; see https://issues.apache.org/jira/browse/SPARK-12269. There's an ongoing discussion in https://github.com/apache/spark/pull/10256 RE: updating Spark's Jackson dependency.

I'm going to close this issue for now, since I don't think there's anything actionable that we can do here from the `spark-redshift` side, but I'd keep an eye on those Spark threads. Thanks for the tip about downgrading the AWS SDK. Let me know if you find a working configuration in your environment.
